### PR TITLE
REP-6194 Added the Python, PIP, and Doc-Utils dependencies to the legacy Jenkins slaves to support the Attribute Mapping library documentation.

### PR DESCRIPTION
**Pull request by wdschei**

### CODE REVIEW

**Comment by dmnjohns** [comment]
It looks like the `attributeMapping` job in Jenkins is labelled `build`, and thus runs on slaves 1-5. If I am reading the Puppets correctly, Slaves 1-5 have the `build_slave` class applied to them rather than the `legacy_slave` class, which means these changes would not apply to the slaves running the `attributeMapping` job.

Am I missing something? What is the reason for not setting up Python and the `doc-utils` library in the `build_slave` class?

**Comment by dmnjohns** [comment]
Also, if we are going to use the `python::pip` class, I would recommend adding `stankevich-python` to `repose_jenkins` module's metadata file. There is a hard dependency there, so adding it as a dependency to the module makes the module more self-contained.

I know Puppet Librarian will handle installing dependency modules for us, but using it as a central source-of-truth rather than appropriately defining our modules' metadata seems questionable. It makes it more difficult to manage conflicts and makes our modules reliant on Librarian, which is not the official Puppet way of managing dependencies.

File: myModules/repose_jenkins/manifests/build_slave.pp [hunk 1]
```diff
@@ -112,9 +112,15 @@
     ensure => present,
   }
 
-  package { 'gcc':
-    # This is needed by API-Checker's use of Nailgun
+  package { [
+    'gcc', # This is needed by API-Checker's use of Nailgun
+    'python-pip', # This is for the Attribute Mapping library documentation
+    'python2.7-dev', # This is for the Attribute Mapping library documentation
```
Comment (dmnjohns) [inline myModules/repose_jenkins/manifests/build_slave.pp]: Another thought. If we are using the Puppet Python module anyway, why bother manually calling out the packages for installation? Though it is unlikely, they could change at some point and break us (although fixing that would probably be the same amount of work as updating the Puppet module).

If we want to use the Puppet module instead, here's how:
https://github.com/stankevich/puppet-python#usage

Comment (wdschei) [reply myModules/repose_jenkins/manifests/build_slave.pp]: Updated.

**Comment by dmnjohns** [comment]
Drats. Running this did not seem to work:
```
root@slave1:~# puppet agent --test
Info: Retrieving pluginfacts
Error: /File[/var/lib/puppet/facts.d]: Could not evaluate: Could not retrieve information from environment production source(s) puppet://puppet/pluginfacts
Info: Retrieving plugin
Info: Loading facts
Info: Caching catalog for slave1.openrepose.org
Warning: /Apt_key[Add key: D05AB914 from Apt::Source rackspace_monitoring]: The id should be a full fingerprint (40 characters), see README.
Info: Applying configuration version '1508266371'
Notice: /Stage[main]/Repose_gradle/Archive[gradle-3.4]/creates: creates changed 'archive not extracted' to 'extracting in /opt to create /opt/gradle-${version}'
Error: Execution of '/usr/bin/apt-get -q -y -o DPkg::Options::=--force-confold install 2.7-dev' returned 100: Reading package lists...
Building dependency tree...
Reading state information...
libpython2.7-dev is already the newest version.
libpython2.7-dev set to manually installed.
python2.7-dev is already the newest version.
python2.7-dev set to manually installed.
The following extra packages will be installed:
  libopenjp2-7
The following NEW packages will be installed:
  libopenjp2-7 libopenjp2-7-dev
0 upgraded, 2 newly installed, 0 to remove and 0 not upgraded.
Need to get 129 kB of archives.
After this operation, 490 kB of additional disk space will be used.
WARNING: The following packages cannot be authenticated!
  libopenjp2-7 libopenjp2-7-dev
E: There are problems and -y was used without --force-yes
Error: /Stage[main]/Python::Install/Package[python-dev]/ensure: change from purged to present failed: Execution of '/usr/bin/apt-get -q -y -o DPkg::Options::=--force-confold install 2.7-dev' returned 100: Reading package lists...
Building dependency tree...
Reading state information...
libpython2.7-dev is already the newest version.
libpython2.7-dev set to manually installed.
python2.7-dev is already the newest version.
python2.7-dev set to manually installed.
The following extra packages will be installed:
  libopenjp2-7
The following NEW packages will be installed:
  libopenjp2-7 libopenjp2-7-dev
0 upgraded, 2 newly installed, 0 to remove and 0 not upgraded.
Need to get 129 kB of archives.
After this operation, 490 kB of additional disk space will be used.
WARNING: The following packages cannot be authenticated!
  libopenjp2-7 libopenjp2-7-dev
E: There are problems and -y was used without --force-yes
Error: Execution of '/usr/bin/apt-get -q -y -o DPkg::Options::=--force-confold install 2.7' returned 100: Reading package lists...
Building dependency tree...
Reading state information...
libpython2.7 is already the newest version.
libpython2.7-dev is already the newest version.
libpython2.7-dev set to manually installed.
libpython2.7-minimal is already the newest version.
libpython2.7-stdlib is already the newest version.
python-crypto is already the newest version.
python-crypto set to manually installed.
python-lxml is already the newest version.
python-ply is already the newest version.
python2.7 is already the newest version.
python2.7-dev is already the newest version.
python2.7-dev set to manually installed.
python2.7-minimal is already the newest version.
python-ldb is already the newest version.
python-ldb set to manually installed.
python-pil is already the newest version.
python-pil set to manually installed.
python-samba is already the newest version.
python-samba set to manually installed.
python-talloc is already the newest version.
python-talloc set to manually installed.
python-tdb is already the newest version.
python-tdb set to manually installed.
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 python-cherrypy3 : Conflicts: python-cherrypy but 2.3.0-4 is to be installed
 python-irc : Conflicts: python-irclib but 0.4.8-1.1 is to be installed
 python-libpcap : Conflicts: python-pypcap but 1.1.2+debian-2.2 is to be installed
 python-nova : Conflicts: python-cjson but 1.1.0-1 is to be installed
 python-pypcap : Conflicts: python-libpcap but 0.6.4-1 is to be installed
 python-quixote1 : Conflicts: python-quixote but 2.7~b2-1 is to be installed
 python-rtslib-fb : Conflicts: python-rtslib but 1:3.0~pre4.1~g1b33ceb-2 is to be installed
 python-xapian1.3 : Conflicts: python-xapian
 python-xattr : Conflicts: python-pyxattr
E: Unable to correct problems, you have held broken packages.
Error: /Stage[main]/Python::Install/Package[python]/ensure: change from purged to present failed: Execution of '/usr/bin/apt-get -q -y -o DPkg::Options::=--force-confold install 2.7' returned 100: Reading package lists...
Building dependency tree...
Reading state information...
libpython2.7 is already the newest version.
libpython2.7-dev is already the newest version.
libpython2.7-dev set to manually installed.
libpython2.7-minimal is already the newest version.
libpython2.7-stdlib is already the newest version.
python-crypto is already the newest version.
python-crypto set to manually installed.
python-lxml is already the newest version.
python-ply is already the newest version.
python2.7 is already the newest version.
python2.7-dev is already the newest version.
python2.7-dev set to manually installed.
python2.7-minimal is already the newest version.
python-ldb is already the newest version.
python-ldb set to manually installed.
python-pil is already the newest version.
python-pil set to manually installed.
python-samba is already the newest version.
python-samba set to manually installed.
python-talloc is already the newest version.
python-talloc set to manually installed.
python-tdb is already the newest version.
python-tdb set to manually installed.
Some packages could not be installed. This may mean that you have
requested an impossible situation or if you are using the unstable
distribution that some required packages have not yet been created
or been moved out of Incoming.
The following information may help to resolve the situation:

The following packages have unmet dependencies:
 python-cherrypy3 : Conflicts: python-cherrypy but 2.3.0-4 is to be installed
 python-irc : Conflicts: python-irclib but 0.4.8-1.1 is to be installed
 python-libpcap : Conflicts: python-pypcap but 1.1.2+debian-2.2 is to be installed
 python-nova : Conflicts: python-cjson but 1.1.0-1 is to be installed
 python-pypcap : Conflicts: python-libpcap but 0.6.4-1 is to be installed
 python-quixote1 : Conflicts: python-quixote but 2.7~b2-1 is to be installed
 python-rtslib-fb : Conflicts: python-rtslib but 1:3.0~pre4.1~g1b33ceb-2 is to be installed
 python-xapian1.3 : Conflicts: python-xapian
 python-xattr : Conflicts: python-pyxattr
E: Unable to correct problems, you have held broken packages.
Notice: /Package[pip]: Dependency Package[python] has failures: true
Warning: /Package[pip]: Skipping because of failed dependencies
Notice: /Package[virtualenv]: Dependency Package[python] has failures: true
Warning: /Package[virtualenv]: Skipping because of failed dependencies
Notice: /Stage[main]/Python/Anchor[python::end]: Dependency Package[python] has failures: true
Notice: /Stage[main]/Python/Anchor[python::end]: Dependency Package[python-dev] has failures: true
Warning: /Stage[main]/Python/Anchor[python::end]: Skipping because of failed dependencies
Notice: /Stage[main]/Repose_jenkins::Build_slave/Python::Pip[doc-utils]/Exec[pip_install_doc-utils]: Dependency Package[python] has failures: true
Notice: /Stage[main]/Repose_jenkins::Build_slave/Python::Pip[doc-utils]/Exec[pip_install_doc-utils]: Dependency Package[python-dev] has failures: true
Warning: /Stage[main]/Repose_jenkins::Build_slave/Python::Pip[doc-utils]/Exec[pip_install_doc-utils]: Skipping because of failed dependencies
Notice: Finished catalog run in 6.14 seconds
```

It appears that, for some reason, the module is not playing nice with our boxes. So maybe your old way is the way to go.

File: myModules/repose_jenkins/manifests/build_slave.pp [hunk 2]
```diff
@@ -117,4 +117,15 @@
...
+  class { 'python' :
+    version    => '2.7',
+    pip        => 'present',
+    dev        => 'present',
+    virtualenv => 'absent',
+    gunicorn   => 'absent',
+  }
+
+  python::pip { 'doc-utils':
+    pkgname => 'doc-utils',
```
Comment (dmnjohns) [inline myModules/repose_jenkins/manifests/build_slave.pp]: A quick `pip install doc-utils` yielded the message:
> Could not find any downloads that satisfy the requirement doc-utils

Perhaps this should be `docutils` instead?

Comment (wdschei) [reply myModules/repose_jenkins/manifests/build_slave.pp]: Updated.

**Pull request merged**